### PR TITLE
impl: benchmark client call round trip

### PR DESCRIPTION
**Pull request by dbolduc**
Motivated by #10407 and other observability efforts.

Add a benchmark to measure the round trip time of a client call (where the server responds instantaneously).

If the cost incurred by logging is < X, we can justify always logging. Especially when compared with the time it takes to serialize/deserialize a message and send it over the wire.

What is X? Probably greater than 1 microsecond.

Also, for the record, the results look different on my laptop. The ratios are very similar though.
```
Run on (8 X 4900 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
Load Average: 0.61, 3.36, 5.01
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_ClientRoundTripStubOnly              1569 ns         1569 ns       372348
BM_ClientRoundTripMetadata              1720 ns         1720 ns       397493
BM_ClientRoundTripLogging               3313 ns         3313 ns       207672
BM_ClientRoundTripTenExtraOptions       3226 ns         3226 ns       215489
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10467)

### CODE REVIEW

File: generator/integration_tests/benchmarks/client_benchmark.cc [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+  struct ExtraOption0 { using Type = int; };
+  struct ExtraOption1 { using Type = int; };
+  struct ExtraOption2 { using Type = int; };
+  struct ExtraOption3 { using Type = int; };
+  struct ExtraOption4 { using Type = int; };
+  struct ExtraOption5 { using Type = int; };
+  struct ExtraOption6 { using Type = int; };
+  struct ExtraOption7 { using Type = int; };
+  struct ExtraOption8 { using Type = int; };
+  struct ExtraOption9 { using Type = int; };
```
Comment (coryan) [inline generator/integration_tests/benchmarks/client_benchmark.cc:216]: Maybe:

```cc
template <int N> struct ExtraOption { using Type = int; };

auto options = Options{}.set<ExtraOption<0>>(0) ... ;
```

File: generator/integration_tests/benchmarks/client_benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+  // clang-format on
+  auto options = Options{}
```
Comment (coryan) [inline generator/integration_tests/benchmarks/client_benchmark.cc:218]: A lot of the options we do use in every call are retry policy and backoff policy.  Those are more expensive to copy than an `int`?

Comment (dbolduc) [reply generator/integration_tests/benchmarks/client_benchmark.cc:218]: Ack. The options are defaulted in `MakeTestConnection()`, so the policies are included in each of these benchmarks..

File: generator/integration_tests/benchmarks/client_benchmark.cc (hunk 1 above)
Comment (dbolduc) [reply generator/integration_tests/benchmarks/client_benchmark.cc]: Done.

File: generator/integration_tests/benchmarks/client_benchmark.cc [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    auto status = client.DoNothing();
+    benchmark::DoNotOptimize(status);
+  }
+}
+BENCHMARK(BM_ClientRoundTripMetadata);
+
+void BM_ClientRoundTripLogging(benchmark::State& state) {
+  auto options = Options{};
+  std::shared_ptr<GoldenKitchenSinkStub> stub = std::make_shared<TestStub>();
+  stub = std::make_shared<GoldenKitchenSinkLogging>(
```
Comment (coryan) [inline generator/integration_tests/benchmarks/client_benchmark.cc:198]: If you want to make the test look better / different, try:

```cc
google::cloud::LogSink::Instance().set_minimum_severity(google::cloud::Severity::GCP_LS_WARNING);
```

**Review by coryan — APPROVED** [review]

File: generator/integration_tests/benchmarks/client_benchmark.cc [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+//   L1 Instruction 32 KiB (x48)
+//   L2 Unified 1024 KiB (x48)
+//   L3 Unified 39424 KiB (x2)
+// Load Average: 0.38, 0.38, 0.73
+// ----------------------------------------------------------------------------
+// Benchmark                                  Time             CPU   Iterations
+// ----------------------------------------------------------------------------
+// BM_ClientRoundTripStubOnly              2330 ns         2329 ns       293640
+// BM_ClientRoundTripMetadata              2693 ns         2693 ns       253897
+// BM_ClientRoundTripLogging               5551 ns         5551 ns       126334
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply generator/integration_tests/benchmarks/client_benchmark.cc:47]: Well that is with logging enabled. The `BM_ClientRoundTripMetadata` is the one that shows the cost of adding one layer of indirection.

I have the same question as you. What evidence will convince us that we can always include the logging/tracing layer. I want to discuss this in the team meeting.

[the remark this answers is not in the document]
Comment (coryan) [reply generator/integration_tests/benchmarks/client_benchmark.cc:47]: * Does the `5551 ns` means that each call takes 5.5 µs (microseconds)?  Or that all 126,334 calls take 5.5 µs?  I reckon the former? 

* Do you know what is the approximate overhead of a single gRPC call?  This suggests it is about 155 µs:

https://grafana-dot-grpc-testing.appspot.com/?orgId=1

* Do you know if the logger was actually enabled, i.e., if  `google::cloud::LogSink::Instance()` had at least one backend and the log severity was enabled?

File: generator/integration_tests/benchmarks/client_benchmark.cc (hunk 3 above)
Comment (dbolduc) [reply generator/integration_tests/benchmarks/client_benchmark.cc:198]:
```
Run on (96 X 2000 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x48)
  L1 Instruction 32 KiB (x48)
  L2 Unified 1024 KiB (x48)
  L3 Unified 39424 KiB (x2)
Load Average: 0.73, 0.29, 0.18
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_ClientRoundTripStubOnly              2353 ns         2353 ns       297597
BM_ClientRoundTripMetadata              2775 ns         2775 ns       253992
BM_ClientRoundTripLogging               2372 ns         2372 ns       293695
BM_ClientRoundTripTenExtraOptions       4987 ns         4986 ns       140084
```

interesting....

**Pull request merged**